### PR TITLE
Update mobile prebuilt package ops to add support for opset 14 and 15

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
Update mobile prebuilt package ops to add support for opset 14 and 15.

Add helper script that can check if an ONNX model would be able to be run using the pre-built package.

**Motivation and Context**
Support newer models in pre-built mobile package.

### CODE REVIEW

File: tools/ci_build/github/android/mobile_package.required_operators.config [hunk 1]
```diff
@@ -9,19 +9,21 @@
 # allow float, int8, uint8. operators that manipulate shapes or indices have int32 and int64 enabled internally.
 !globally_allowed_types;float,int8_t,uint8_t
 
-# ops used by the tf2onnx tflite converter. same list for opsets 12 and 13.
+# ops used by the tf2onnx tflite converter. full list for opset 12. incremental list for later opsets.
```
Comment (skottmckay) [inline tools/ci_build/github/android/mobile_package.required_operators.config:12]: Made the updates incremental so it's easier to see how much changed and have a rough idea of binary size impact.

File: tools/python/check_model_can_use_ort_mobile_pkg.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+import argparse
+import onnx
+import os
+import pathlib
+import sys
+from util import reduced_build_config_parser
+
+
+def check_graph(graph, opsets, required_ops, global_types, unsupported_ops):
```
Comment (edgchen1) [inline tools/python/check_model_can_use_ort_mobile_pkg.py:15]: so we assume that:
- globally allowed types are specified
- only model inputs need to be checked and not intermediate values

is that generally obvious or would it be worth documenting?

File: tools/python/check_model_can_use_ort_mobile_pkg.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+                opset not in required_ops[domain] or \
+                node.op_type not in required_ops[domain][opset]:
+            unsupported_ops.add(f'{domain}:{opset}:{node.op_type}')
+
+        # recurse into subgraph for control flow nodes (Scan/Loop/If)
+        for attr in node.attribute:
+            if attr.HasField('g'):
+                check_graph(attr.g, opsets, required_ops, global_types, unsupported_ops)
+
+    return
```
Comment (edgchen1) [inline tools/python/check_model_can_use_ort_mobile_pkg.py:48]: >return [](http://example.com/codeflow?start=4&length=6)

nit: don't need trailing return

File: tools/ci_build/github/android/mobile_package.required_operators.config [hunk 4]
```diff
@@ -9,19 +9,30 @@
...
+# ops used by the tf2onnx tflite converter. 
+ai.onnx;12,13,14,15;Abs,Add,And,ArgMax,ArgMin,AveragePool,Cast,Ceil,Clip,Concat,ConstantOfShape,Conv,ConvTranspose,Cos,CumSum,DepthToSpace,DequantizeLinear,Div,DynamicQuantizeLinear,Elu,Equal,Exp,Expand,Flatten,Floor,Gather,GatherND,Gemm,Greater,GreaterOrEqual,Identity,If,LRN,LeakyRelu,Less,LessOrEqual,Log,LogSoftmax,Loop,MatMul,Max,MaxPool,Mean,Min,Mul,Neg,NonMaxSuppression,NonZero,Not,Or,PRelu,Pad,Pow,QuantizeLinear,Range,Reciprocal,ReduceMax,ReduceMean,ReduceMin,ReduceProd,ReduceSum,Relu,Reshape,Resize,ReverseSequence,Round,ScatterND,Shape,Sigmoid,Sin,Size,Slice,Softmax,SpaceToDepth,Split,Sqrt,Squeeze,Sub,Sum,Tanh,ThresholdedRelu,Tile,TopK,Transpose,Unique,Unsqueeze,Where
```
Comment (edgchen1) [inline tools/ci_build/github/android/mobile_package.required_operators.config:13]: >12,13,14,15 [](http://example.com/codeflow?start=8&length=11)

nice update to support multiple opsets in a line :)

File: tools/python/check_model_can_use_ort_mobile_pkg.py (hunk 2 above)
Comment (skottmckay) [reply tools/python/check_model_can_use_ort_mobile_pkg.py:15]: We use the global type list in the pre-built package config so I think it's fair to assume that. 

Checking more things gets much harder. As it's the ONNX model there's no type info for intermediate values in the model. Additionally it gets pretty nuanced with things like the int32/int64 support being a required type on specific inputs/outputs of specific operators. I didn't want to re-create all that logic at this point as I'm not sure how frequently a model not being able to be run in the pre-built package is due to the inner types used. Mainly trying to have an ability to quickly assess an ONNX model (e.g. we just exported from PT) being likely to be able to run with the pre-built package. 

I'll see if there are some other simple checks to add without going overboard.

**Comment by gwang-msft** [comment]
In the future, can we make this part of the ort model conversion script, which can give warning at the time of conversion? (no need to clone the repo)

**Review by gwang-msft — DISMISSED** [review]

**Comment by skottmckay** [comment]
> In the future, can we make this part of the ort model conversion script, which can give warning at the time of conversion? (no need to clone the repo)

Yes. Want to get it a little more refined and stable first and after that I can move it into utils dir so it's part of the package. Also want something to roughly estimate the number of partitions if CoreML or NNAPI are enabled to recommend whether full and/or basic optimization should be tried.

File: tools/python/check_model_can_use_ort_mobile_pkg.py (hunk 2 above)
Comment (skottmckay) [reply tools/python/check_model_can_use_ort_mobile_pkg.py]: I added a few more checks and tested against all the models used in the generation of the pre-built package config. With the change there are no more false negatives, and the messaging calls out that it's not a complete check.

File: tools/python/check_model_can_use_ort_mobile_pkg.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    'BFloat16': 16
+}
+
+tensorproto_type_to_cpp = {v: k for k, v in cpp_to_tensorproto_type.items()}
+
+
+def check_graph(graph, opsets, required_ops, global_types, special_types, unsupported_ops):
+    '''
+    Check the graph and any subgraphs for usage of types or operators which we know are not supported.
+    There are some gray areas around operators using int32/int64/bools which
```
Comment (edgchen1) [inline tools/python/check_model_can_use_ort_mobile_pkg.py:40]: >which [](http://example.com/codeflow?start=71&length=5)

is there more to this sentence?

File: tools/python/check_model_can_use_ort_mobile_pkg.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+    for t in global_cpp_types:
+        if t in cpp_to_tensorproto_type:
+            global_onnx_tensorproto_types.add(cpp_to_tensorproto_type[t])
+        else:
+            print(f'Error: Unexpected data type of {t}')
+            sys.exit(-1)
+
+    # a subset of operators require int32 and int64 to always be enabled as they're used with shapes and indices.
+    # additional we have a number of operators (e.g. Not) that use bool which isn't included in type reduction.
```
Comment (edgchen1) [inline tools/python/check_model_can_use_ort_mobile_pkg.py:144]: >additional [](http://example.com/codeflow?start=6&length=10)

nit: "additionally"

File: tools/python/check_model_can_use_ort_mobile_pkg.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description='Check if model is likely to be able to be run using the ONNX Runtime Mobile Pre-Built Package',
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+
+    script_dir = os.path.dirname(os.path.realpath(__file__))
+
+    # config that was used to create the pre-build package.
```
Comment (edgchen1) [inline tools/python/check_model_can_use_ort_mobile_pkg.py:161]: >build [](http://example.com/codeflow?start=45&length=5)

"pre-built"

**Review by edgchen1 — DISMISSED** [review]
:shipit:

**Review by edgchen1 — APPROVED** [review]
:shipit:

**Pull request merged**